### PR TITLE
Initialize any_nan_ property of LinearTreeLearner

### PR DESCRIPTION
**Pull request by btrotta**
I think this will fix one of the valgrind issues found in #3699

### CODE REVIEW

**Review by btrotta — COMMENTED** [review]
/gha run r-valgrind

**Review by guolinke — APPROVED** [review]

File: src/treelearner/linear_tree_learner.cpp [hunk 1]
```diff
@@ -34,6 +34,7 @@ void LinearTreeLearner::InitLinear(const Dataset* train_data, const int max_leav
       }
     }
   }
+  any_nan_ = false;
```
Comment (StrikerRUS) [inline src/treelearner/linear_tree_learner.cpp:37]: /gha run r-valgrind

Comment (StrikerRUS) [reply src/treelearner/linear_tree_learner.cpp:37]: ✔️ 

```
==1916== 
==1916== HEAP SUMMARY:
==1916==     in use at exit: 308,761,984 bytes in 55,584 blocks
==1916==   total heap usage: 2,772,225 allocs, 2,716,641 frees, 6,118,912,031 bytes allocated
==1916== 
==1916== 336 bytes in 1 blocks are possibly lost in loss record 152 of 2,682
==1916==    at 0x483DD99: calloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==1916==    by 0x40149CA: allocate_dtv (dl-tls.c:286)
==1916==    by 0x40149CA: _dl_allocate_tls (dl-tls.c:532)
==1916==    by 0x5706322: allocate_stack (allocatestack.c:622)
==1916==    by 0x5706322: pthread_create@@GLIBC_2.2.5 (pthread_create.c:660)
==1916==    by 0x56D4DDA: ??? (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
==1916==    by 0x56CC8E0: GOMP_parallel (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
==1916==    by 0x154C1F21: LGBM_DatasetCreateFromCSC (c_api.cpp:1305)
==1916==    by 0x154F10B0: LGBM_DatasetCreateFromCSC_R (lightgbm_R.cpp:91)
==1916==    by 0x4942146: R_doDotCall (dotcode.c:634)
==1916==    by 0x494CFDD: do_dotcall (dotcode.c:1281)
==1916==    by 0x49A06DF: bcEval (eval.c:7072)
==1916==    by 0x498C2C0: Rf_eval (eval.c:727)
==1916==    by 0x498F02F: R_execClosure (eval.c:1897)
==1916== 
==1916== LEAK SUMMARY:
==1916==    definitely lost: 0 bytes in 0 blocks
==1916==    indirectly lost: 0 bytes in 0 blocks
==1916==      possibly lost: 336 bytes in 1 blocks
==1916==    still reachable: 308,761,648 bytes in 55,583 blocks
==1916==                       of which reachable via heuristic:
==1916==                         newarray           : 4,264 bytes in 1 blocks
==1916==         suppressed: 0 bytes in 0 blocks
==1916== Reachable blocks (those to which a pointer was found) are not shown.
==1916== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==1916== 
==1916== For lists of detected and suppressed errors, rerun with: -s
==1916== ERROR SUMMARY: 3 errors from 2 contexts (suppressed: 0 from 0)
writing valgrind output to valgrind-logs.log
valgrind found 0 bytes definitely lost
valgrind found 0 bytes indirectly lost
valgrind found 336 bytes possibly lost
```
https://github.com/microsoft/LightGBM/runs/1680889485?check_suite_focus=true

Comment (StrikerRUS) [reply src/treelearner/linear_tree_learner.cpp:37]: cc @jameslamb

Comment (jameslamb) [reply src/treelearner/linear_tree_learner.cpp:37]: I'm glad it worked but that doesn't tell us that this PR fixed the issues I saw in #3699. Without #3699 , none of the tests in the R package cover the linear learner stuff, and so valgrind isn't actually checking this code.

**Pull request merged**